core/src/main/java/org/bitcoinj/core/SegwitAddress.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2018 Andreas Schildbach
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bitcoinj.core;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+
+import javax.annotation.Nullable;
+
+import org.bitcoinj.params.Networks;
+
+public class SegwitAddress extends AbstractAddress {
+    public SegwitAddress(NetworkParameters params, int witnessVersion, byte[] witnessProgram)
+            throws AddressFormatException {
+        this(params, encode(witnessVersion, witnessProgram));
+    }
+
+    private static byte[] encode(int witnessVersion, byte[] witnessProgram) throws AddressFormatException {
+        try {
+            ByteArrayOutputStream enc = new ByteArrayOutputStream(64 + 1);
+            enc.write(witnessVersion);
+            enc.write(convertBits(witnessProgram, 0, witnessProgram.length, 8, 5, true));
+            return enc.toByteArray();
+        } catch (IOException x) {
+            throw new RuntimeException(x);
+        }
+    }
+
+    private SegwitAddress(NetworkParameters params, byte[] bytes) {
+        super(params, bytes);
+        if (bytes.length < 1)
+            throw new AddressFormatException("Zero data found");
+        final int witnessVersion = witnessVersion();
+        if (witnessVersion > 16)
+            throw new AddressFormatException("Invalid script version");
+        byte[] witnessProgram = witnessProgram();
+        if (witnessProgram.length < 2 || witnessProgram.length > 40)
+            throw new AddressFormatException("Invalid length");
+        // check script length for version 0
+        if (witnessVersion == 0 && witnessProgram.length != 20 && witnessProgram.length != 32)
+            throw new AddressFormatException("Invalid length for address version 0");
+    }
+
+    private int witnessVersion() {
+        return bytes[0] & 0xff;
+    }
+
+    private byte[] witnessProgram() {
+        // skip version byte
+        return convertBits(bytes, 1, bytes.length - 1, 5, 8, false);
```

Comment:
We can avoid this conversion if we saved the witness version and program in `VersionedChecksummedBytes`.

Reply:
We could transiently cache them in `SegwitAddress` if needed.